mlflow/pyspark/ml/__init__.py

```diff
@@ -0,0 +1,282 @@
+    safe_patch,
+    try_mlflow_log,
+)
+
+
+_logger = logging.getLogger(__name__)
+
+
+def _get_fully_qualified_class_name(instance):
+    return instance.__class__.__module__ + "." + instance.__class__.__name__
+
+
+def _read_log_model_allowlist_from_file(allowlist_file):
+    allowlist = set()
+    with open(allowlist_file) as f:
+        for line in f:
+            stripped = line.strip()
+            is_blankline_or_comment = stripped == "" or stripped.startswith("#")
+            if not is_blankline_or_comment:
+                allowlist.add(stripped)
+    return allowlist
+
+
+def _read_log_model_allowlist(spark_session):
+    """
+    Reads the module allowlist and returns it as a set.
+    """
+    allowlist_file = spark_session.sparkContext._conf.get(
+        "spark.databricks.mlflow.autolog.log_model.allowlistFile", None
+    )
+    builtin_allowlist_file = resource_filename(__name__, "log_model_allowlist.txt")
+    if allowlist_file:
+        try:
+            return _read_log_model_allowlist_from_file(allowlist_file)
+        except Exception:
+            # fallback to built-in allowlist file
+            _logger.warning(
+                "Reading config spark.databricks.mlflow.autolog.log_model.allowlistFile "
+                f"{allowlist_file} failed, fallback to built-in allowlist file"
+            )
+            return _read_log_model_allowlist_from_file(builtin_allowlist_file)
+    else:
+        return _read_log_model_allowlist_from_file(builtin_allowlist_file)
+
+
+# When autolog called, `_log_model_allowlist` will be initialized.
+_log_model_allowlist = None
+
+
+def _get_warning_msg_for_skip_log_model(model):
+    return (
+        f"This model {model.uid} is not logged because it is not in "
+        "allowlist or its nested models are not in allowlist."
+    )
+
+
+def _should_log_model(spark_model):
+    # TODO: Handle PipelineModel/CrossValidatorModel/TrainValidationSplitModel
+    class_name = _get_fully_qualified_class_name(spark_model)
+    if class_name in _log_model_allowlist:
+        if class_name == "pyspark.ml.classification.OneVsRestModel":
+            return _should_log_model(spark_model.models[0])
+        else:
+            return True
+    else:
+        return False
+
+
+def _get_estimator_info_tags(estimator):
+    """
+    :return: A dictionary of MLflow run tag keys and values
+             describing the specified estimator.
+    """
+    return {
+        "estimator_name": estimator.__class__.__name__,
+        "estimator_class": _get_fully_qualified_class_name(estimator),
+    }
+
+
+def _get_instance_param_map(instance):
+    from pyspark.ml.param import Params
+
+    param_map = {
+        param.name: instance.getOrDefault(param)
+        for param in instance.params
+        if instance.isDefined(param)
+    }
+    # TODO:
+    #  handle special case: CrossValidator.estimatorParamMaps
+
+    expanded_param_map = {}
+    for k, v in param_map.items():
+        if isinstance(v, Params):
```

Comment:
Could you add a code comment to clarify when a value in `param_map` could itself be an instance of `Params`? e.g. does this happen for meta-estimators like Pipelines?

Reply:
This case happens in param e.g. `OneVsRest.classifier`/`CrossValidator.estimator`, this case param value is an Estimator instance (inherits `pyspark.ml.param.Params`)